---
title: "Untitled"
author: "Nechama Wieder"
date: "2022-12-05"
output: html_document
---
5'UTR Analysis across LEOUF deciles
low LOEUF = low tolerance to inactivation ie more likely to be disease causing, also haploinsufficient
high LOEUF = high tolerance to inactivation ie less likely to be disease causing

#Data gnomAD V2
downloaded from https://gnomad.broadinstitute.org/downloads >> Under heading of “Constraint” >> “pLoF Metrics by Transcript”, file = "gnomad.v2.1.1.lof_metrics.by_transcript.txt.bgz"

#Analysis Content
1. 5'UTR Length
2. 5'UTR Introns
3. uORF/oORF/start-stops
4. uORF analysis
  4.1 uORF length
  4.2 uORF to CDS distance
5. uAUG/length metric
6. Ribo-seq uORFs
  . uORFs codon optimality
  . uORF lengths
  . uORF start codons
7. Translational Efficiency (Norderer)
8. PhyloP (5UTR, uORF, Start-stop)
9. 5'UTR GC content
10. CAGE 5'UTR Peaks
11. Minimum Free Energy (MFE) scores 
12. CADD


Load Libraries
```{r message=FALSE, error=FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)

```

#Prepare gnomAD Data 
```{r message=FALSE, error=FALSE, warning=FALSE}
gnomad1 <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/3.  5 UTR in disease genes/gene subsets databases/gnomad.v2.1.1.lof_metrics.by_transcript.txt.bgz", fill = TRUE, header = TRUE)

#from nicky
#LOEUF has been calculated for all transcripts in GRCh37. MANE is on GRCh38 and hence many MANE transcripts are not represented in the LOEUF file. To get around this, we can take a single LOEUF score per gene - easiest way to do this is to take the value assigned to the canonical transcript (marked in the file) for every gene. Canonical transcripts are not perfect, this is why we use MANE wherever we can, but for this purpose they are a good proxy per gene.

#restrict to canonical, rebin (make new deciles), keep highest decile where genes are duplicated (ie >1 transcript for a gene as i want 1 trans: 1 gene)
gnomad <- gnomad1 %>% filter(canonical == "true")
gnomad <- gnomad[!is.na(gnomad$oe_lof_upper), ] #remove those with oe_lof_upper = na
#how many unique gene names are there?
#un1 <- unique(gnomad$gene) #19155, 42  genes entered >1
#for the genes >1, take the lower oe_lof_upper
gnomad_2 <- gnomad %>% group_by(gene) %>% slice(which.min(oe_lof_upper))#19155

### rebin unique canonical transcripts
gnomad_2$decile <- ntile(gnomad_2$oe_lof_upper, 10)

#subset gn_final to just the ones i want:
gn_final1 <- subset(gnomad_2, select=c("gene", "transcript", "pLI", "decile", "oe_lof_upper", "oe_lof_upper_bin", "oe_lof_upper_rank", "no_lofs", "obs_lof","transcript_type"))
```
#1. 5'UTR Lengths 
```{r message=FALSE, error=FALSE, warning=FALSE}
#bring in MANE 5' UTR length
length_exon <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/length_introns_12.12.22.txt")

#####merge by gene id called transcript type in gnomad data
length_exon <- cSplit(length_exon, "gene_id", ".") 
names(gn_final1)[10]<-"gene_id"
names(length_exon)[6]<-"gene_id"
gnomad_len <- merge(length_exon, gn_final1, by = "gene_id") 
#17337

####stats - length differences in bottom and top 2 deciles

bot <- gnomad_len %>% filter(decile<3)
mean(bot$UTR_length)#268.8319

top <- gnomad_len %>% filter(decile>8)
mean(top$UTR_length)#162.3923
bot2 <- bot$UTR_length
top2 <- top$UTR_length
wilcox.test(bot2, top2)$p.value
#p=8.965308e-145
#significant

####plot
gnomad_len$decile <- gnomad_len$decile*10

ggplot(gnomad_len, aes(x=factor(decile), y=UTR_length)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("UTR Length (bp)") + geom_hline(yintercept = 202,linetype = "dotted", size=0.8) + theme(text = element_text(size = 23)) + scale_y_continuous(limits = c(0, 1500))

#nicky said maybe truncate at 2000. ylim may not be good so try + scale_y_continuous(limits = c(0, 1500))

# this colour is the 5utr colour from phylop/cadd plot #793A92

#some journals might not want boxplot so try make raincloud with dots over a boxplot alongside

#https://ourcodingclub.github.io/tutorials/dataviz-beautification/

library(tidyverse)
library(ggthemes)  # for a mapping theme
library(ggalt)  # for custom map projections
library(ggrepel)  # for annotations
library(viridis)  # for nice colours

#remotes::install_github('jorvlan/raincloudplots')
#library(raincloudplots)

#source("https://gist.githubusercontent.com/benmarwick/2a1bb0133ff568cbe28d/raw/fb53bd97121f7f9ce947837ef1a4c65a73bffb3f/geom_flat_violin.R") - geom_flat_violin

#need to play with it more and se

ggplot(data = gnomad_len, aes(x=factor(decile), y = UTR_length, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = UTR_length, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="#793A92") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.8, fill="#793A92") +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") +  scale_y_continuous(limits = c(0, 2000)) + guides(fill = FALSE, color = FALSE) +theme_light()

#dont love all purple, how about the dots a grey?
#here dots are grey and boxplot purple. if keep this boxplot its nice on the purple dot background
ggplot(data = gnomad_len, aes(x=factor(decile), y = UTR_length, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = UTR_length, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="grey48") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.5, fill="#793A92") +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") + coord_cartesian(ylim=c(0, 2000)) + guides(fill = FALSE, color = FALSE) +theme_light()  + theme(text = element_text(size = 20)) 


#how about violin plot - my current fave. think its neatest but doesnt show individual data points whihc is waht htey want
ggplot(data = gnomad_len, aes(x=factor(decile), y = UTR_length, fill = decile)) + geom_violin(fill="#793A92")  +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") +  scale_y_continuous(limits = c(0, 1500)) + guides(fill = FALSE, color = FALSE) +theme_light() + geom_boxplot(width=0.1, color="grey", alpha=0.1) 


#i dont think the truncation is wrong - look carefully at the individual points plotted


#try stat halfeye again
ggplot(gnomad_len, aes(x = factor(decile), y = UTR_length))  + ggdist::stat_halfeye(aes(fill=decile, fill = after_scale(colorspace::lighten(fill, .4))))+ theme_light() + theme(legend.position = "none") 




#ggplot(data = gnomad_len, aes(x=factor(decile), y = UTR_length, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8) +     geom_point(aes(y = UTR_length, color = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1) +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.8) +     labs(x = "LEOUF Decile", y = "UTR Length (bp)") +     guides(fill = none, color = none)   +  scale_y_continuous(limits = c(0, 2000)) 
    # Picking nicer colours
    scale_fill_manual(values = c("#793A92")) 
    scale_colour_manual(values = c("#793A92")) + theme_light()



 
    



# was truncating oddly, so need to try again
#nicky said its fine to have boxplot on the one side, so try truncate at 2000bps (make sure its doing it right) and change colour..
 

```
#2. 5'UTR Introns 
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. plot diff between 0 and 1+ introns across LEOUF
gn_int<- gnomad_len %>% mutate(intron_cutoff = intron_final>0)
gn_int2 <- gn_int %>% group_by(decile) %>% dplyr::count(intron_cutoff)

total_tab <- as.data.frame(table(gnomad_len$decile))
total <- total_tab$Freq
#get percentages
tot <- rep(total, each=2)
gn_int2$total <- tot
gn_int2$perc <- (gn_int2$n/gn_int2$total) *100

#prepare for plot
#change intron true/false into meaningful
gn_int2 <- gn_int2 %>% mutate(intron_cutoff=str_replace(intron_cutoff, "TRUE", "1+ Intron"))
gn_int2 <- gn_int2 %>% mutate(intron_cutoff=str_replace(intron_cutoff, "FALSE", "0 Intron"))

ggplot(gn_int2, aes(x=factor(decile), y = perc, fill = intron_cutoff)) + geom_bar(position="dodge", stat="identity") + theme_light()   + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("0 Intron" = "#BFACC8", "1+ Intron" = "#2B4595")) + theme(legend.title=element_blank()) + theme(text = element_text(size = 20))  #+ theme(legend.position="none") 



#2. Are the differences between 0 intron and >1 intron across deciles statistically significant?
#####stats
#make matrix
zero <- gn_int2 %>% filter(intron_cutoff=="0 Intron")
int <- gn_int2 %>% filter(intron_cutoff=="1+ Intron")

bot_z <- zero %>% filter(decile<30)
botz_genes <- sum(bot_z$n)
botz_tot <- sum(bot_z$total)

top_z <- zero %>% filter(decile>80)
top_zgenes <- sum(top_z$n)
topz_tot <- sum(top_z$total)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene <- c(botz_tot, topz_tot)
#how many genes in those deciles with 1+ intron
one_int <- c(botz_genes, top_zgenes)

int_matrix <- data.frame(zero="",int=one_int, total=tot_dec_gene)
int_matrix$zero <- int_matrix$total - int_matrix$int
#get percentages so can report in paper
int_matrix$perc_z <- (int_matrix$zero / int_matrix$total) *100
int_matrix$perc_o <- (int_matrix$int / int_matrix$total) *100
#do chi-square
test_int <- subset(int_matrix, select = c(zero, int))
chisq.test(test_int)$p.value

#0.185136

```
#3. uAUG/uORF/oORF/Start-stops
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. get orf per gene data 
orf_gene <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/orf_gene 14.12.22.txt")

#get gene_id
orf2 <- merge(orf_gene, length_exon, by="transcript")


#2. merge with gnomad
orf_gn <- merge(gn_final1, orf2, by="gene_id")

#3. aggregate to get totals of each uaug type per decile
orf_gn2 <- orf_gn %>% group_by(decile,orf_type) %>% summarise(n()) 
#n here is number of genes with at least 1 of that orf type


#3. get total per decile and append
#total <- data.frame(table(orf_gn$decile))
#this total is gnomad genes with uorfs but i want all gnomad genes

#use gnomad length for all totals - these are all gnomad genes which have a 5utr in mane
total <- data.frame(table(gnomad_len$decile))
totals <- total$Freq
totals2 <- rep(totals, each=3)
orf_gn2$total <- totals2

#percentage
orf_gn2$perc <- (orf_gn2$`n()`/orf_gn2$total) *100

#change all start-stop to Start-Stop
orf_gn2$orf_type <- as.character(orf_gn2$orf_type)
orf_gn2[orf_gn2 == "start-stop"] <- "Start-Stop"

###plot

orf_gn2$orf_type <- factor(orf_gn2$orf_type, levels = c("uORF", "oORF", "Start-Stop"))
orf_gn2$decile <- orf_gn2$decile*10 

#add mean lines 
table(orf_gene$orf_type)
#uo - 6461/18764 == 34.4%
#oorf - 2806/18764 == 15%
#ss - 940/18764 == 5%

#old version
#ggplot(orf_gn2, aes(x=factor(decile), y=perc, fill=orf_type)) + geom_bar(position="dodge", stat="identity") + theme_light() + xlab("LOEUF Decile (%)") + ylab("Percentage") + scale_fill_manual(values = c("uORF" = "NAVY", "oORF" = "palevioletred1", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept = 34.4, colour="NAVY",linetype = "dotted") + geom_hline(yintercept = 15, colour="palevioletred1",linetype = "dotted") + geom_hline(yintercept =5, colour="deepskyblue1",linetype = "dotted") # + theme(legend.title=element_blank()) + theme(legend.position = "none")  

ggplot(orf_gn2, aes(x=factor(decile), y=perc, fill=orf_type)) + geom_bar(position="dodge", stat="identity") + theme_light() + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("uORF" = "NAVY", "oORF" = "palevioletred1", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept = 34.4, colour="NAVY",linetype = "dotted") + geom_hline(yintercept = 15, colour="palevioletred1",linetype = "dotted") + geom_hline(yintercept =5, colour="deepskyblue1",linetype = "dotted")  + theme(legend.title=element_blank()) + theme(text = element_text(size = 20)) #+ theme(legend.position = "none")  

#########STATS
####uORFs
uo <- orf_gn2 %>% filter(orf_type=="uORF")
###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 uorfs
bot_uo <-uo %>% filter(decile<30) 
top_uo <-  uo %>% filter(decile>80) 

#total of genes in bot and top 2 deciles
tot_bot_u <- sum(bot_uo$total)#3695
tot_top_u <- sum(top_uo$total)#2789

#total genes with uo in bot and top
botuo <- sum(bot_uo$`n()`)
topuo<- sum(top_uo$`n()`)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene_u <- c(tot_bot_u, tot_top_u)
#how many genes in those deciles with 1+ uo
one_uo <- c(botuo, topuo)

uo_matrix <- data.frame(zero="",uo=one_uo, total=tot_dec_gene_u)
uo_matrix$zero <- uo_matrix$total - uo_matrix$uo
#get percentages so can report in paper
uo_matrix$perc_z <- (uo_matrix$zero / uo_matrix$total) *100
uo_matrix$perc_o <- (uo_matrix$uo / uo_matrix$total) *100
#do chi-square
test_uo <- subset(uo_matrix, select = c(zero, uo))
chisq.test(test_uo)$p.value
#p=2.087572e-51
#sig




####start-stops
ss <- orf_gn2 %>% filter(orf_type=="Start-Stop")
###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 start stops
bot_ss <- ss %>% filter(decile<30) #sum(bot_ss$`n()`) == 252 genes in bottom 2 deciles have at least 1 start-stops
top_ss <-  ss %>% filter(decile>80) #sum(top_ss$`n()`) 125 genes in bottom 2 deciles have at least 1 start-stops

#total of genes in bot and top 2 deciles
tot_bot <- sum(bot_ss$total)#3653
tot_top <- sum(top_ss$total)#2864

#total genes with ss in bot and top
botss <- sum(bot_ss$`n()`)
topss<- sum(top_ss$`n()`)

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene <- c(tot_bot, tot_top)
#how many genes in those deciles with 1+ ss
one <- c(botss, topss)

ss_matrix <- data.frame(zero="",ss=one, total=tot_dec_gene)
ss_matrix$zero <- ss_matrix$total - ss_matrix$ss
#get percentages so can report in paper
ss_matrix$perc_z <- (ss_matrix$zero / ss_matrix$total) *100
ss_matrix$perc_o <- (ss_matrix$ss / ss_matrix$total) *100
#do chi-square
test <- subset(ss_matrix, select = c(zero, ss))
chisq.test(test)$p.value
#p=8.506704e-05
#sig

```
#4. uORF analysis (length and cds-is) 
```{r message=FALSE, error=FALSE, warning=FALSE}
######4.1 uORF length
#bring in data
uo_len <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/uo_ 20.9.22.txt")
uo_len <- merge(uo_len, length_exon, by="transcript")

uorflen <- merge(gn_final1, uo_len, by="gene_id")
uorflen$decile <- uorflen$decile*10

mean_uole <- mean(uo_len$uorf_len)
ggplot(uorflen, aes(x=factor(decile), y=uorf_len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("uORF Length (bp)") + geom_hline(yintercept = mean_uole, linetype = "dotted", size=0.8)+ theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, 600))


#stats
bot_len <- uorflen %>% filter(decile<30)
#mean
mean(bot_len$uorf_len)#52.46849
top_len <- uorflen %>% filter(decile>80)
mean(top_len$uorf_len)
#59.13161
botl <- bot_len$uorf_len
topl <- top_len$uorf_len

wilcox.test(botl, topl)$p.value

#p=4.315421e-06





######4.2 uORF to CDS distance
#last uORF-CDS distance
#bring in data
uORF_laststop <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/uorfs reinitiation dis 20.9.22.txt")

uORF_laststop <- merge(uORF_laststop, length_exon, by="transcript")

uorf_only <- uORF_laststop %>% filter(orf_type=="uORF")#6320
#merge with gn_final1
gn_dis <- merge(gn_final1, uorf_only, by="gene_id")

#plot dis by decile
gn_dis$decile <- gn_dis$decile*10

mean_dis <- mean(uorf_only$dis)

ggplot(gn_dis, aes(x=factor(decile), y=dis)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("uORF to CDS distance (bp)")  + geom_hline(yintercept = mean_dis,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))  + coord_cartesian(ylim=c(0, 750))


##stats
bot_dec2 <- gn_dis %>% filter(decile<30)
mean(bot_dec2$dis) #98.83082
top_dec2 <- gn_dis %>% filter(decile>80)
mean(top_dec2$dis) # 76.87811

bot2 <- bot_dec2$dis
top2 <- top_dec2$dis
wilcox.test(bot2, top2)$p.value
#p= 0.0001266438
```
#5. uAUG/length metric
```{r message=FALSE, error=FALSE, warning=FALSE}
atg_length <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/uaug_metric_m1.txt")

#atg_length

names(atg_length)[5]<- "gene_id"
gn_metric <- merge(gnomad_len, atg_length, by = "gene_id")#7375

#plot metric range in boxplots
gn_metric$decile <- gn_metric$decile *10

#add mean line
mean_met <- mean(atg_length$metric)# 0.01040724

ggplot(gn_metric, aes(x=factor(decile), y=metric)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8)  + theme_light() + xlab("LOEUF Decile") + ylab("uAUG/length") + geom_hline(yintercept = mean_met,linetype = "dotted") + theme(text = element_text(size = 20))



#####stats
#top and bottom 2 deciles
top_met <- gn_metric %>% filter(decile>80)
bot_met<- gn_metric %>% filter(decile<30)

#means
mean(top_met$metric) #0.01292813 == 0.013
mean(bot_met$metric) #0.00845449 == 0.009

topmet <- top_met$metric
botmet <- bot_met$metric

wilcox.test(botmet, topmet,)$p.value
#2.774926e-58

```
#6. Ribo-seq uORFs 
```{r message=FALSE, error=FALSE, warning=FALSE}
###1. diff between 0, 1 and 1+ ribo-seq uorfs across deciles
###2. codon optimality of riboseq uorfs across deciles
###3. ribo-seq uorfs start codons across deciles
###4. ribo-seq uorf lengths across deciles


###1. diff between 0, 1 and 1+ ribo-seq uorfs across deciles
#bring in data
smorf_gene <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/smorf_gene.txt")

sm_gn <- merge(gn_final1, smorf_gene, by="gene_id")#3513 

#
no_smorf <- anti_join(gn_final1, smorf_gene, by = "gene_id")#15642
one_smorf <- sm_gn %>% filter(number_uORF == 1) #2574
more_smorf <- sm_gn %>% filter(number_uORF > 1) #939

#group each into deciles
no_smorf_tab <- data.frame(table(no_smorf$decile))
one_smorf_tab <- data.frame(table(one_smorf$decile))
more_smorf_tab <- data.frame(table(more_smorf$decile))

#total in each decile
total_per_decile <- data.frame(table(gn_final1$decile))

no_smorf_tab$total <- total_per_decile$Freq
one_smorf_tab$total <- total_per_decile$Freq
more_smorf_tab$total <- total_per_decile$Freq

#add group name
no_smorf_tab$group <- "0"
one_smorf_tab$group <- "1"
more_smorf_tab$group <- ">1"

#percentage
no_smorf_tab$perc <- no_smorf_tab$Freq/no_smorf_tab$total *100
one_smorf_tab$perc <- one_smorf_tab$Freq/one_smorf_tab$total *100
more_smorf_tab$perc <- more_smorf_tab$Freq/more_smorf_tab$total *100


#rbind to plot
smorf_plot <- rbind(no_smorf_tab, one_smorf_tab, more_smorf_tab)
#reorder bars
smorf_plot$group <- factor(smorf_plot$group, levels = c("0", "1", ">1"))

#change deciles to be multiples of 10
smorf_plot$Var1 <- as.numeric(smorf_plot$Var1)
smorf_plot$Var1 <- smorf_plot$Var1 *10

#plot
ggplot(smorf_plot, aes(x = factor(Var1), y = perc, fill = group )) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage") + guides(fill=guide_legend(title="Ribo-seq uORF Number")) + scale_fill_manual(values = c("0" = "#BFACC8", "1" = "#2B4595", ">1" = "#228C80"))+ theme(text = element_text(size = 20))  +# theme(legend.position="none")


#stats
#difference between 0 and 1+ uorfs in bot vs top 2 deciles

###stats for bottom and top 2 deciles start-stops between ahving 0 or 1 uorfs
bot_smrf <-smorf_plot %>% filter(Var1<30) 
top_smrf <-  smorf_plot %>% filter(Var1>80) 

#total of genes in bot and top 2 deciles

#group them as 0 or 1 uorf
bot_smrf[bot_smrf==">1"]<-1
#sum all those in bttom deciles

bot1 <- bot_smrf %>% group_by(group) %>% summarise(Freq=sum(Freq))

top_smrf[top_smrf==">1"]<-1
top1 <- top_smrf %>% group_by(group) %>% summarise(Freq=sum(Freq))

statsuo <- merge(bot1, top1, by="group")
statsuo <- subset(statsuo, select=-c(group))

chisq.test(statsuo)$p.value
#1.692544e-141

#make df with all this in one go
#how many genes in the bottom and top deciles
tot_dec_gene_u <- c(tot_bot_u, tot_top_u)
#how many genes in those deciles with 1+ uo
one_uo <- c(botuo, topuo)

uo_matrix <- data.frame(zero="",uo=one_uo, total=tot_dec_gene_u)
uo_matrix$zero <- uo_matrix$total - uo_matrix$uo
#get percentages so can report in paper
uo_matrix$perc_z <- (uo_matrix$zero / uo_matrix$total) *100
uo_matrix$perc_o <- (uo_matrix$uo / uo_matrix$total) *100
#do chi-square
test_uo <- subset(uo_matrix, select = c(zero, uo))
chisq.test(test_uo)$p.value
#p=2.087572e-51
#sig





###2. codon optimality of riboseq uorfs across deciles
#bring in optimality data
ribo_opt2 <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/smorf uorf codon optimality HeLa 28.7.22.txt")
 
gn_rib <- merge(gn_final1, ribo_opt2,by="gene_id")#4834

#plot
gn_rib$decile <- gn_rib$decile *10

mean_co <- mean(ribo_opt2$score)
ggplot(gn_rib, aes(x=factor(decile), y=score)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light()  + xlab("LOEUF Decile") + ylab("Codon Optimality Score") + geom_hline(yintercept = mean_co, linetype = "dotted", size=0.8) + theme(text = element_text(size = 20)) 

#stats
bot_co <- gn_rib %>% filter(decile<30)
top_co <- gn_rib %>% filter(decile>80)

botco <- bot_co$score
topco <- top_co$score
wilcox.test(botco, topco)
#p-value = 0.1688

###3. ribo-seq uorfs start codons across deciles
#from smorf2 data
uORFs <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/ribo_seq_uo.txt")
#merge with gnomad
gn_smorf <- merge(gn_final1, uORFs,by="gene_id")

#i want to plot atg vs not atg across loeuf
gn_smorf$start_type <- "non-canonical"

#if ATG change to "canonical"
for(l in 1:nrow(gn_smorf)){
  if (gn_smorf$starts[l] == "ATG"){
    gn_smorf$start_type[l] <- (gn_smorf$start_type[l] <- "canonical")
  }
}

#aggregate start type by decile
#tab_starts <- aggregate(gn_smorf$start_type, by = list(gn_smorf$decile), FUN=sum)#not working
starts_tab <- gn_smorf %>% group_by(decile, start_type)%>%summarise(n())
starts_tab2 <- gn_smorf %>% group_by(decile)%>%summarise(n()) #how many in each decile
#append total to starts_tab
dec_tot <- starts_tab2$`n()`
dec_tot2 <- rep(dec_tot, each=2)
#get percentages
starts_tab$total <- dec_tot2
starts_tab$perc <- starts_tab$`n()`/starts_tab$total *100

#plot
starts_tab$decile <- starts_tab$decile*10
starts_tab$start_type <- factor(starts_tab$start_type, levels=c("canonical","non-canonical"))

ggplot(starts_tab, aes(x=factor(decile), y=perc, fill=start_type)) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")  + guides(fill=guide_legend(title="Start Type")) + scale_fill_manual(values = c("canonical" = "#BFACC8", "non-canonical" = "#2B4595")) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank())  #+ theme(legend.position="none")


#in black and grey
ggplot(starts_tab, aes(x=factor(decile), y=perc, fill=start_type)) + geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")  + guides(fill=guide_legend(title="Start Type")) + scale_fill_manual(values = c("canonical" = "grey", "non-canonical" = "black")) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank())  #+ theme(legend.position="none")

#stats
names(starts_tab)[3]<-"n"

bot_start <- starts_tab %>% filter(decile<30)
top_start <- starts_tab %>% filter(decile>80)

#group by and sum 
bot <- bot_start %>% group_by(start_type) %>% summarise(n=sum(n))
top <- top_start %>% group_by(start_type) %>% summarise(n=sum(n))
names(bot)[2]<-"bot"
names(top)[2]<-"top"

can_stats<- merge(bot, top, by="start_type")

#transpose?
cs2 <- as.data.frame(t(can_stats)) 
#remove top line
cs2 <- cs2[-1,]

cs2$V1 <- as.numeric(cs2$V1)
cs2$V2 <- as.numeric(cs2$V2)
chisq.test(cs2)$p.value
#0.183124





###4. ribo-seq uorf lengths across deciles
#gn_smorf has length in it (col:len)
gn_smorf$decile <- gn_smorf$decile*10

mean_rs <- mean(uORFs$len)
ggplot(gn_smorf, aes(x=factor(decile), y=len)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("Ribo-seq uORF Length (bp)") + geom_hline(yintercept = mean_rs, linetype = "dotted", size=0.8) + theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, 600))

#stats
bot_rs <- gn_smorf %>% filter(decile<30)
top_rs <- gn_smorf %>% filter(decile>80)

botrs <- bot_rs$len
toprs <- top_rs$len

wilcox.test(botrs, toprs)
#p=0.9453
```
#7. Translational Efficiency (Noderer) 
```{r message=FALSE, error=FALSE, warning=FALSE}
#from TE for uAUG markdown - ORF_TE (/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/1. 5 UTR features/TE for uAUGs)
#want boxplot of TE for each disease set
#bring in TE data
ORF_te2 <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/orf_TE_noderer_m1.txt")
#think this is called ORF_te2 now
#names(ORF_te2)[5]<- "transcript"
te <- merge(ORF_te2, length_exon, by="transcript")


TE_gnomad <- merge(gn_final1, te, by="gene_id")

#just uorfs
te_uo <- TE_gnomad %>% filter(orf_type=="uORF")
mean_te <- mean(te_uo$efficiency)#78.77569

te_uo$decile <- te_uo$decile*10

ggplot(te_uo, aes(x=factor(decile), y=efficiency)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("TE (Norderer)") + geom_hline(yintercept = mean_te,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))

#stats
bot_te <- te_uo %>% filter(decile<30)
top_te <- te_uo %>% filter(decile>80)

bote <- bot_te$efficiency
tope <- top_te$efficiency

wilcox.test(bote, tope)$p.value
# 0.583095

```
#8. PhyloP 
```{r message=FALSE, error=FALSE, warning=FALSE}
#1. bring in PhyloP data
####5'UTR mean PhyloP
phylop <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/five_utr_ph_m1_10.1.23.txt")
phylop <- cSplit(phylop, "gene_id", ".")
names(phylop)[7]<-"gene_id"
#merge
utr_ph <- merge(phylop, gn_final1,by="gene_id")

####uORF (start and ends) and start-stops
phy2 <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/uorf_ss phylop m1 10.1.23.txt")
#merge with mane to get gene names
names(phy2)[1]<-"transcript"
ph_name <- merge(length_exon, phy2, by="transcript")

#merge gnomad
ph_gn <- merge(gn_final1, ph_name, by="gene_id")

########plot mean 5'UTR, uorf start, uorf end and start stops
#2. combine to make one df
names(utr_ph)[3]<-"phylop"
utr_ph <- subset(utr_ph, select=c(gene.x, gene_id, phylop,decile))
utr_ph$group <- "5'UTR"
utr_ph$decile <- utr_ph$decile*10

#append to ph_gn
ph_gn$decile<-ph_gn$decile*10
#subset
ph_gn2 <- subset(ph_gn, select=c(gene.x, gene_id, phylop,decile, group))

plot_ph <- rbind(ph_gn2, utr_ph)
 #change all uorf end to uorf stop
plot_ph[plot_ph=="uORF end"]<- "uORF Stop"

#plot
#order
plot_ph$group <- factor(plot_ph$group, levels = c("5'UTR", "uORF Start", "uORF Stop", "Start-Stop"))

#old version
#ggplot(plot_ph, aes(x=factor(decile), y=phylop, fill=group)) + geom_boxplot() + theme_light() + xlab("LOEUF Decile") + ylab("PhyloP") + scale_fill_manual(values = c("5'UTR" = "#793A92", "uORF Start" = "navy", "uORF Stop" = "navy", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept =2,colour="blue",linetype = "dotted", size=0.6) # + theme(legend.title=element_blank()) + theme(legend.position = "none") 

ggplot(plot_ph, aes(x=factor(decile), y=phylop, fill=group)) + geom_boxplot() + theme_light() + xlab("LOEUF Decile (%)") + ylab("PhyloP") + scale_fill_manual(values = c("5'UTR" = "#793A92", "uORF Start" = "navy", "uORF Stop" = "navy", "Start-Stop" = "deepskyblue1"))  + geom_hline(yintercept =2,linetype = "dotted", size=0.6) + theme(text = element_text(size = 20))  + theme(legend.title=element_blank()) #+ theme(legend.position = "none") 

#save plot for ease when want to fiddle with colours
#write.table(plot_ph, "/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/loeuf phylop plot 10.1.23.txt")



############STUDENT T TEST
#for 5UTR mean phylop
#nicky said why dont we do student t test?
#1.2 top and bottom 2 deciles
bot2 <- utr_ph %>% filter(decile<30)

x <- sum(bot2$phylop[which(bot2$phylop>=0)])
y <- sum(bot2$phylop[which(bot2$phylop<=0)])
z <- x+y #(x-y added it)
#get number of values in bot
z/nrow(bot2) 
#mean = 1.101596

top2 <- utr_ph %>% filter(decile>80)

x <- sum(top2$phylop[which(top2$phylop>=0)])
y <- sum(top2$phylop[which(top2$phylop<=0)])
z <- x+y #(x-y added it)
#get number of values in bot
z/nrow(top2) 
#0.01249374

#students t-test
t.test(bot2$phylop, top2$phylop)
#p-value < 2.2e-16
#but when i add $p.value = 0 and when i put the test into a variable it says p val = 0

library(tidyverse)
library(ggpubr)
library(rstatix)


bt_ph <- bot2$phylop
tp_ph <- top2$phylop
t.test(bt_ph, tp_ph)$p.value

#all giving 0

#stats for uorf start, ends and start-stops also needed
bot <- ph_gn2 %>% filter(decile<30)
top <- ph_gn2 %>% filter(decile>80)

###uorf start
bot_uos <- bot %>% filter(group=="uORF Start")
top_uos <- top %>% filter(group=="uORF Start")

botuos <- bot_uos$phylop
topuos <- top_uos$phylop

t.test(botuos,topuos)$p.value
#p-value < 2.2e-16

####uorf end
bot_uoe <- bot %>% filter(group=="uORF end")
top_uoe <- top %>% filter(group=="uORF end")

botuoe <- bot_uoe$phylop
topuoe <- top_uoe$phylop

t.test(botuoe,topuoe)$p.value
#p-value < 2.2e-16

####start-stops
bot_ss <- bot %>% filter(group=="Start-Stop")
top_ss <- top %>% filter(group=="Start-Stop")

botss <- bot_ss$phylop
topss <- top_ss$phylop

t.test(botss,topss)$p.value
#p-value < 2.2e-16

```
#9.GC content
```{r message=FALSE, error=FALSE, warning=FALSE}
gc_save<- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/man1_gc.txt")

gc_save <- merge(length_exon, gc_save, by="transcript")

gc_gnomad <- merge(gn_final1, gc_save, by="gene_id")

#plot
gc_gnomad$decile <- gc_gnomad$decile*10

#add mean intercept
mean_gc <- mean(gc_save$gc_percent)

ggplot(gc_gnomad, aes(x=factor(decile), y=gc_percent)) + geom_boxplot(aes(group = decile), fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("GC Content per bp (%)") + geom_hline(yintercept = mean_gc,linetype = "dotted", size=0.8) + theme(text = element_text(size = 20))


###stats
#chi square on bottom and top 2 deciles GC content

bot_gc <- gc_gnomad %>% filter(decile<30) 
mean(bot_gc$gc_percent)#67.3%
top_gc <-  gc_gnomad %>% filter(decile>80)
mean(top_gc$gc_percent)#60%

bot_gc <- bot_gc$gc_percent
top_gc <- top_gc$gc_percent

wilcox.test(bot_gc, top_gc)$p.value
#4.946138e-134

#do stats on when u restrict it to 100-300bps
mid<- gc_gnomad %>% filter(utr_len %in% 100:300) #6514

bot_gc1 <- mid %>% filter(decile<30) 
mean(bot_gc1$gc_percent)#69%
top_gc1 <-  mid %>% filter(decile>80)
mean(top_gc1$gc_percent)#60%

bot_gc1 <- bot_gc1$gc_percent
top_gc1 <- top_gc1$gc_percent

wilcox.test(bot_gc1, top_gc1)$p.value
# 2.350705e-89


```
#10.CAGE
```{r message=FALSE, error=FALSE, warning=FALSE}
cage_count <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/cage m1 13.12.22.txt")
cage_count <- cSplit(cage_count, "gene_id", ".")
names(cage_count)[6]<-"gene_id"
cage_gn <- merge(gn_final1, cage_count, by="gene_id")#16233

#add groupings for how i want to plot..
#gna do 1, 2-5, 6+

cage_one <- cage_gn %>% filter(count <2)#2528
cage_mid <- cage_gn%>% filter(count >= 2 & count <= 5)#8536
cage_six <- cage_gn %>% filter(count >5)#4807

cage_one$group <- "1"
cage_mid$group <- "2-5"
cage_six$group <- "≥6"

cage_pl <- rbind(cage_one, cage_mid, cage_six)
cage_pl2 <- cage_pl %>% group_by(decile, group) %>% summarise(n())

#add total col
total <- data.frame(table(cage_gn$decile))
totals2 <- rep(total$Freq, each=3)
cage_pl2$total <- totals2

#percentage
cage_pl2$perc <- cage_pl2$`n()`/cage_pl2$total *100
cage_pl2$decile <-cage_pl2$decile*10

#plot
cage_pl2$group <- factor(cage_pl2$group, levels=c("1", "2-5", "≥6") )

#mean cage peak


ggplot(cage_pl2, aes(x=factor(decile), y=perc, fill=group))+ geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")+ guides(fill=guide_legend(title="CAGE Peaks")) + scale_fill_manual(values = c("1" = "#2B4595", "2-5" = "#BFACC8", "≥6" = "#228C80")) + theme(text = element_text(size = 15)) #+ theme(legend.position="none") 

#change colours  to 3 shades of green
ggplot(cage_pl2, aes(x=factor(decile), y=perc, fill=group))+ geom_bar(position="dodge", stat="identity") + theme_light()  + xlab("LOEUF Decile") + ylab("Percentage")+ guides(fill=guide_legend(title="CAGE Peaks")) + scale_fill_manual(values = c("1" = "#7CDFD4", "2-5" = "#228C80", "≥6" = "#14524B")) + theme(text = element_text(size = 20)) #+ theme(legend.position="none") 

#+ theme(text = element_text(size = 15)) for leg

#its not coming up the more than sign when exporting plot so manially added it in ppt



###stats###
#showing that more peaks in lowest 2 dec vs highest
#diff between 1 and 1+ peaks?


bot_cage <- cage_gn %>% filter(decile<3)
top_cage <- cage_gn %>% filter(decile>8)


#bot 2 deciles
bot_cage <-bot_cage%>% mutate(one_cutoff = count>1)#differentiate above and below 1
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "1+"))#change to meaningful vals
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "1"))
total <- nrow(bot_cage) 
b <- data.frame(table(bot_cage$one_cutoff))
b$total <- total
b$perc <- (b$Freq/b$total)*100
#transpose
b3 <- as.data.frame(t(b))
#select the only row I want
b4<- b3 %>% slice(2)
names(b4)[1]<-"1"
names(b4)[2]<-"1+"
rownames(b4) <- c("bot")
b4$`1` <- as.numeric(b4$`1`)
b4$`1+` <- as.numeric(b4$`1+`)
#round
b4<- b4 %>% mutate_if(is.numeric, round)

#top 2 deciles

top_cage <-top_cage%>% mutate(one_cutoff = count>1)#differentiate above and below 1
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "1+"))#change to meaningful vals
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "1"))
total2 <- nrow(top_cage) 
t <- data.frame(table(top_cage$one_cutoff))
t$total <- total2
t$perc <- (t$Freq/t$total)*100
#transpose
df3 <- as.data.frame(t(t))
#select the only row I want
df4<- df3 %>% slice(2)
names(df4)[1]<-"1"
names(df4)[2]<-"1+"
rownames(df4) <- c("top")
df4$`1` <- as.numeric(df4$`1`)
df4$`1+` <- as.numeric(df4$`1+`)
#round
df4<- df4 %>% mutate_if(is.numeric, round)

#bind
cage_stat <- rbind(b4, df4)

chisq.test(cage_stat)$p.value
#1.518424e-89

###stats for > or equal 6 peaks
bot_cage <-bot_cage%>% mutate(one_cutoff = count>=6)#differentiate above and below 1
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "6+"))#change to meaningful vals
bot_cage <- bot_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "<6"))
total <- nrow(bot_cage) 
b <- data.frame(table(bot_cage$one_cutoff))
b$total <- total
b$perc <- (b$Freq/b$total)*100
#transpose
b3 <- as.data.frame(t(b))
#select the only row I want
b4<- b3 %>% slice(2)
names(b4)[1]<-"<6"
names(b4)[2]<-"6+"
rownames(b4) <- c("bot")
b4$`<6` <- as.numeric(b4$`<6`)
b4$`6+` <- as.numeric(b4$`6+`)
#round
b4<- b4 %>% mutate_if(is.numeric, round)

#top 2 deciles

top_cage <-top_cage%>% mutate(one_cutoff = count>=6)#differentiate above and below 1
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "TRUE", "6+"))#change to meaningful vals
top_cage <- top_cage %>% mutate(one_cutoff=str_replace(one_cutoff, "FALSE", "<6"))
total2 <- nrow(top_cage) 
t <- data.frame(table(top_cage$one_cutoff))
t$total <- total2
t$perc <- (t$Freq/t$total)*100
#transpose
df3 <- as.data.frame(t(t))
#select the only row I want
df4<- df3 %>% slice(2)
names(df4)[1]<-"<6"
names(df4)[2]<-"6+"
rownames(df4) <- c("top")
df4$`<6` <- as.numeric(df4$`<6`)
df4$`6+` <- as.numeric(df4$`6+`)
#round
df4<- df4 %>% mutate_if(is.numeric, round)

#bind
cage_stat <- rbind(b4, df4)

chisq.test(cage_stat)$p.value
#8.059105e-111
```
#11. MFE
```{r message=FALSE, error=FALSE, warning=FALSE}
#go through original and copt all over
energy_score <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/freefold energyscore m1 13.12.22.txt")

energy_score <- merge(length_exon, energy_score, by="transcript")

#merge with gnomad, boxplot across deciles
gn_energy <- merge(gn_final1, energy_score, by="gene_id")

#plot
gn_energy$decile <- gn_energy$decile*10
#ggplot(gn_energy, aes(x=factor(decile), y=score_met)) + geom_boxplot(colour="navy") + theme_light() + xlab("LOEUF Decile") + ylab("Minimum Folding Energy") +ggtitle("5'UTR Minimum Free Folding Energy Metric")
#not sure about dividing by length

#plot the raw scores too
#add mean energy for all genes
mean(energy_score$energy) #-78.81686

#old format
#ggplot(gn_energy, aes(x=factor(decile), y=energy)) + geom_boxplot(fill="#BFACC8") + theme_light() + xlab("LOEUF Decile") + ylab("Minimum Folding Energy")  + geom_hline(yintercept = -78.8, colour="blue",linetype = "dotted", size=0.8) + ylim(-1000, 0)

ggplot(gn_energy, aes(x=factor(decile), y=energy)) + geom_boxplot(fill="#793A92", alpha=0.8) + theme_light() + xlab("LOEUF Decile") + ylab("Minimum Folding Energy")  + geom_hline(yintercept = -78.8, linetype = "dotted", size=0.8) + ylim(-1000, 0) + theme(text = element_text(size = 23)) 

#try raincloud plot
ggplot(data = gn_energy, aes(x=factor(decile), y = energy, fill = decile)) +   geom_flat_violin(position = position_nudge(x = 0.2, y = 0), alpha = 0.8, fill="#793A92") +     geom_point(aes(y = energy, fill = decile), position = position_jitter(width = 0.15), size = 1, alpha = 0.1, color="grey48") +     geom_boxplot(width = 0.2, outlier.shape = NA, alpha = 0.5, fill="#793A92") +     labs(x = "LEOUF Decile", y = "Minimum Folding Energy") + guides(fill = FALSE, color = FALSE) +theme_light() + theme(text = element_text(size = 20)) + coord_cartesian(ylim=c(0, -1000))

#Warning messages:
#1: Removed 6 rows containing non-finite values (`stat_ydensity()`). 
#2: Removed 6 rows containing non-finite values (`stat_boxplot()`). 
#3: Removed 265 rows containing missing values (`geom_point()`).
#not sure whats it removing. how i find teh missing vals? the normla orig boxlpot didnt have this...



####stats
bot_mfe <- gn_energy %>% filter(decile<30)
mean(bot_mfe$energy)#-115.1429
top_mfe <- gn_energy %>% filter(decile>80)
mean(top_mfe$energy)#-54.80042

#wilcoxon
botmfe <- bot_mfe$energy
topmfe <- top_mfe$energy
wilcox.test(botmfe, topmfe)$p.value
#7.349057e-206


#do stats on when u restrict it to 100-300bps
mid<- gn_energy %>% filter(utr_len %in% 100:300) #6514

bot2 <- mid %>% filter(decile<30)
mean(bot2$energy)#-81.52125
top2 <-mid %>% filter(decile>80)
mean(top2$energy)#-60.01709

#wilcoxon
bot_2 <- bot2$energy
top_2 <- top2$energy
wilcox.test(bot_2, top_2)$p.value
# 2.976304e-57

```
#12. CADD 
```{r message=FALSE, error=FALSE, warning=FALSE}
#bring in CADD data
cadd <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/cadd_all_m1_17.12.22.txt")
names(cadd)[1] <-"transcript"
cadd2 <- merge(cadd, length_exon, by="transcript")


#merge with gnomad
gn_cadd <- merge(gn_final1, cadd2, by="gene_id")

#plot
gn_cadd$decile <- gn_cadd$decile*10


gn_cadd$group <- factor(gn_cadd$group, levels = c("5'UTR", "uORF Start", "uORF Stop", "Start-Stop"))


ggplot(gn_cadd, aes(x=factor(decile), y=cadd, fill=group)) + geom_boxplot() + theme_light() + xlab("LOEUF Decile") + ylab("CADD") + scale_fill_manual(values = c("5'UTR" = "#793A92", "uORF Start" = "navy", "uORF Stop" = "navy", "Start-Stop" = "deepskyblue1")) + theme(text = element_text(size = 20)) # + theme(legend.position="none")  
#stats
#mean of 5'UTR CADD bottom and top deciles
cad_utr <- gn_cadd %>% filter(group=="5'UTR")
#function
means <- function(df){
  x <- sum(df$cadd[which(df$cadd>=0)])
y <- sum(df$cadd[which(df$cadd<=0)])
z <- x+y #(x-y added it)
#get number of values in df
return(z/nrow(df))
}

#comparing top and bottom 2 deciles
bot2 <- cad_utr %>% filter(decile<30)
means(bot2)
#1.483858
top2<- cad_utr %>% filter(decile>80)
means(top2)
#0.6942674

#stats
#5utr cadd
t.test(bot2$cadd, top2$cadd)
#p-value < 2.2e-16

#uo start
cad_uosta <- gn_cadd %>% filter(group=="uORF Start")
bot_u <- cad_uosta %>% filter(decile<30)
top_u <-cad_uosta %>% filter(decile>80)

t.test(bot_u$cadd, top_u$cadd)
#p-value < 2.2e-16

#uo stop
cad_uostp <- gn_cadd %>% filter(group=="uORF Stop")
bot_u2 <- cad_uostp %>% filter(decile<30)
top_u2 <-cad_uostp %>% filter(decile>80)

t.test(bot_u2$cadd, top_u2$cadd)
#p-value < 2.2e-16

#ss
cad_ss <- gn_cadd %>% filter(group=="Start-Stop")
bot_ss <- cad_ss %>% filter(decile<30)
top_ss <-cad_ss %>% filter(decile>80)

t.test(bot_ss$cadd, top_ss$cadd)
#p-value < 2.2e-16



```
#extra analysis
#13. uAUG Consensus sequences
```{r}
#compare the uAUG consensus sequences for top and bottom 2 LEOUF deciles
#bring in all uaugs
uo_ss <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/uo_ss_12.12.22.txt")
oorf <- read.table("/Users/nwieder/OneDrive - Nexus365/NWIEDER DPhil Projects/1_5'UTR_summary_stats/MANE Analysis/5' UTR analysis/5UTr paper/code for paper/MANE v1/oorf_12.12.22.txt")

#put into 1 table
oorf$orf_type <- "oORF"

uoss2 <- subset(uo_ss, select=-c(stop_pos))
oorf2 <- subset(oorf, select=-c(utr_len, oorfend, three, integer, frame, oORF_seq))
aug <- rbind(uoss2, oorf2)

#get range of sequence around aug (excludes those v close to CDS and utr start)
minus <- aug %>% filter(atg_pos>15)
#get+15
minus$atg_end <- minus$atg_pos+2
minus$utr_len <- str_length(minus$sequence)
minus$end_dis <- minus$utr_len - minus$atg_end
plus <- minus %>% filter(end_dis>15)
plus$minus <- str_sub(plus$sequence, start = plus$atg_pos-15, end =plus$atg_pos-1) 
plus$plus <-  str_sub(plus$sequence, start=plus$atg_end+1, end= plus$atg_end+15)
plus$plot <- paste(plus$minus, plus$plus, sep="XXX")

require(ggplot2)
require(ggseqlogo)

#merge with gnomad
names(gnomad_len)[3]<-"transcript"
gn_koz <- merge(gnomad_len, plus, by="transcript")

bot_koz <- gn_koz %>% filter(decile<30)
top_koz <- gn_koz %>% filter(decile>80)

#plot bottom 2 deciles
ggplot() + geom_logo(bot_koz$plot) + theme_logo() +ggtitle("bottom louef uaugs") + ylab("Information Content")
#plot top 2 deciles
ggplot() + geom_logo(top_koz$plot) + theme_logo() +ggtitle("top louef uaugs") + ylab("Information Content")
```
#14. leaderless mRNAs across LEOUF
```{r}
#from mane file df "no_utr"
#prepare no_utr
no_utr <- cSplit(no_utr, "gene_id", ".")
names(no_utr)[14]<-"gene_id"
names(gn_final1)[10]<-"gene_id"

#merge
gn_noutr <- merge(gn_final1, no_utr, by="gene_id")#166 (so missing about 100)
#298 total in mane 1
#plot
gn_noutr$decile <- gn_noutr$decile*10
gn_no_tab <- gn_noutr %>% group_by(decile) %>% summarise(n())

ggplot(gn_no_tab, aes(x=factor(decile), y = `n()`)) + geom_bar(position="dodge", stat="identity") + theme_light()   + xlab("LOEUF Decile") + ylab("Percentage") + scale_fill_manual(values = c("0 Intron" = "#BFACC8", "1+ Intron" = "#2B4595")) + theme(legend.title=element_blank()) + theme(text = element_text(size = 20))  #+ theme(legend.position="none") 

#total_dec <- gn_final1 %>% group_by(decile) %>% summarise(n())
#total_dec <- total_dec$`n()`

#top 2 loeuf deciles
73+40=113
113 of 166
#68.1%

#what kind of genes?
or <- gn_noutr %>% filter(grepl("OR", symbol))#107
#how many in deciles
test <- or %>% group_by(decile) %>% summarise(n())
